application_sdk/activities/.cursor/BUGBOT.md

```diff
@@ -0,0 +1,424 @@
+# Activity Code Review Guidelines - Temporal Activities
+
+## Context-Specific Patterns
+
+This directory contains Temporal activity implementations that perform the actual work of workflows. Activities handle external I/O, database operations, and non-deterministic tasks.
+
+### Phase 1: Critical Activity Safety Issues
+
+**External Resource Safety:**
+
+- All external connections (database, API, file) must have explicit timeouts
+- Connection failures must be handled gracefully with proper retry logic
+- Resource cleanup must happen in finally blocks or context managers
+- Sensitive data must not be logged or exposed in error messages
+- All user inputs must be validated before processing
+
+**Activity Timeout Management:**
+
+- Activities must respect Temporal heartbeat timeouts for long-running operations
+- Progress should be reported via heartbeat for operations > 30 seconds
+- Activities should check for cancellation requests periodically
+- Timeout values must be realistic for the operation being performed
+
+```python
+# ✅ DO: Proper activity with heartbeat and cancellation
+@activity.defn
+async def process_large_dataset_activity(dataset_config: dict) -> dict:
+    total_records = await get_record_count(dataset_config)
+    processed = 0
+
+    async for batch in process_in_batches(dataset_config):
+        # Check for cancellation
+        activity.heartbeat({"progress": processed, "total": total_records})
+
+        try:
+            await process_batch(batch)
+            processed += len(batch)
+        except Exception as e:
+            activity.logger.error(f"Batch processing failed: {e}", exc_info=True)
+            raise
+
+    return {"processed_records": processed}
```

Comment:
we have this decorator called `@auto_heartbeater` and prefer that over a manual heartbeat.